### PR TITLE
NestedDiff

### LINKED ISSUE #14: Nested Diff/Support for sections in UIKit utils
Diffing trees is a hard problem and the best algorithms out there take `O(n^3)` time to compute.
However diffing content separated into sections is a very well defined problem and it doesn't require anything near as complicated as tree diffing.

A function with this signature would be enough:
```swift
func nestedDiff<T: Collection>(from: T, to: T) -> NestedDiff 
where T.Iterator.Element: Collection, 
T.Iterator.Element.Iterator.Element: Equatable
````

1. The function flatMaps the input and output and memorises the indices at which the separate collections start/end. 
2. a normal diff is performed on the flatMapped collection
3. it is mapped back into the nested diff.

When this is implemented it will be easy to add an overload on the UIKit utility functions. if the arguments to `animateRowChanges` are nested collections it will to perform the animations.

### PR DESCRIPTION
**Pull request by wokalski**
This PR fixes #14 
 
- [x] basic support for sections
- [x] shifting elements if there are empty sections
- [x] shifting elements when there are different number of sections
- [x] report deleted sections as a whole
- [x] readme
- [x] documentation
- [x] example
- [x] more tests

### CODE REVIEW

File: DiffTests/NestedDiffTests.swift [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * same
+ * different number of sections
+ */
+class NestedDiffTests: XCTestCase {
+
+    func testDiffOutputs() {
+
+        let expectations: [([[Int]], [[Int]], String)] = [
+            ([], [[1, 2], [1]], "I(0,0)I(1,0)I(0,1)"),
+            ([[1, 2], []], [], "D(0, 0)D(1, 0)"),
```
Comment (farzadshbfn) [inline DiffTests/NestedDiffTests.swift]: Shouldn't the order of deletes be reversed? If it's not patched, it will cause run-time error.

Comment (wokalski) [reply DiffTests/NestedDiffTests.swift]: It's a `Diff`, not `Patch`. See [here](https://github.com/wokalski/Diff.swift#diff-vs-patch-sorting) for more info.

**Comment by wokalski** [comment]
`NestedDiff` is implemented.
I'd appreciate a review but will merge it tomorrow if there's no reviewer.

**Comment by wokalski** [comment]
hey @harlanhaskins. The compiler segfaults and some point and I am not sure how to debug/work around it. Do you have any clues?

**Comment by wokalski** [comment]
I should mention that it doesn't segfault if we compile using `-O -whole-module-optimization`

**Pull request merged**